api/sessionrecording/session_recording.go

```diff
@@ -0,0 +1,298 @@
+	// is an extra precaution to avoid
+	// accidental endless loop due to logic error crashing the system
+	// and allows ctx timeout to kick in if specified
+	var checkpointIteration int64
+	for {
+		checkpointIteration++
+		if checkpointIteration%maxIterationLimit == 0 {
+			select {
+			case <-ctx.Done():
+				if ctx.Err() != nil {
+					return nil, trace.Wrap(ctx.Err())
+				}
+				return nil, trace.LimitExceeded("context has been canceled")
+			default:
+			}
+		}
+		switch r.state {
+		case protoReaderStateEOF:
+			return nil, io.EOF
+		case protoReaderStateError:
+			return nil, r.error
+		case protoReaderStateInit:
+			// read the part header that consists of the protocol version
+			// and the part size (for the V1 version of the protocol)
+			_, err := io.ReadFull(r.reader, r.sizeBytes[:Int64Size])
+			if err != nil {
+				// reached the end of the stream
+				if errors.Is(err, io.EOF) {
+					r.state = protoReaderStateEOF
+					return nil, err
+				}
+				return nil, r.setError(trace.ConvertSystemError(err))
+			}
+			protocolVersion := binary.BigEndian.Uint64(r.sizeBytes[:Int64Size])
+			if protocolVersion != ProtoStreamV1 {
+				return nil, trace.BadParameter("unsupported protocol version %v", protocolVersion)
+			}
+			// read size of this gzipped part as encoded by V1 protocol version
+			_, err = io.ReadFull(r.reader, r.sizeBytes[:Int64Size])
+			if err != nil {
+				return nil, r.setError(trace.ConvertSystemError(err))
+			}
+			partSize := binary.BigEndian.Uint64(r.sizeBytes[:Int64Size])
+			// read padding size (could be 0)
+			_, err = io.ReadFull(r.reader, r.sizeBytes[:Int64Size])
+			if err != nil {
+				return nil, r.setError(trace.ConvertSystemError(err))
+			}
+			r.padding = int64(binary.BigEndian.Uint64(r.sizeBytes[:Int64Size]))
+			gzipReader, err := newGzipReader(io.NopCloser(io.LimitReader(r.reader, int64(partSize))))
+			if err != nil {
+				return nil, r.setError(trace.Wrap(err))
+			}
+			r.gzipReader = gzipReader
+			r.state = protoReaderStateCurrent
+			continue
+			// read the next version from the gzip reader
+		case protoReaderStateCurrent:
+			// the record consists of length of the protobuf encoded
+			// message and the message itself
+			_, err := io.ReadFull(r.gzipReader, r.sizeBytes[:Int32Size])
+			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					return nil, r.setError(trace.ConvertSystemError(err))
+				}
+
+				// due to a bug in older versions of teleport it was possible that padding
+				// bytes would end up inside of the gzip section of the archive. we should
+				// skip any dangling data in the gzip secion.
+				n, err := io.CopyBuffer(io.Discard, r.gzipReader.inner, r.messageBytes[:])
+				if err != nil {
+					return nil, r.setError(trace.ConvertSystemError(err))
+				}
+
+				if n != 0 {
+					// log the number of bytes that were skipped
+					slog.DebugContext(ctx, "skipped dangling data in session recording section", "length", n)
```

Comment:
I wonder if skipped byes is something that should be tracked in `ReaderStats`.

Reply:
I like the idea. I've added a `SkippedBytes` to `ReaderStats`.